sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala

```diff
@@ -520,9 +524,33 @@ trait ColumnResolutionHelper extends Logging with DataTypeErrorsBase {
             "planId" -> planId.toString,
             "q" -> q.toString))
       }
-    })
 
-    val isMetadataAccess = u.getTagValue(LogicalPlan.IS_METADATA_COL).isDefined
+      val matched = found.flatMap { case (child, nodes) =>
+        val resolved = nodes
+          .flatMap(resolveUnresolvedAttributeByPlan(u, _, isMetadataAccess))
+        if (isMetadataAccess) {
+          // NOTE: A metadata column might appear in `output` instead of `metadataOutput`.
+          val metadataOutputSet = child.outputSet ++ AttributeSet(child.metadataOutput)
+          resolved.filter(metadataOutputSet.contains)
+        } else {
+          resolved.filter(child.outputSet.contains)
```

Comment:
`resolveUnresolvedAttributeByPlanId` is not recursive, which means we only do this pruning once at the very end. Shall we make `resolveUnresolvedAttributeByPlanId` recursive to deal with cases that have many joins in one query?

Reply:
The goal is to do this pruning as early as possible.